File: 2296.design-a-text-editor.java
#
# @lc app=leetcode id=2296 lang=java
#
# [2296] Design a Text Editor
#

# @lc code=start
class TextEditor {

    private StringBuilder left;
    private StringBuilder right;

    public TextEditor() {
        left = new StringBuilder();
        right = new StringBuilder();
    }

    private String getLeft10() {
        int sz = left.length();
        if (sz == 0) {
            return "";
        }
        return sz <= 10 ? left.toString() : left.substring(sz - 10);
    }

    public void addText(String text) {
        left.append(text);
    }

    public int deleteText(int k) {
        int del = Math.min(k, left.length());
        left.delete(left.length() - del, left.length());
        return del;
    }

    public String cursorLeft(int k) {
        for (int i = 0; i < k; ++i) {
            if (left.length() == 0) {
                break;
            }
            char c = left.charAt(left.length() - 1);
            left.deleteCharAt(left.length() - 1);
            right.append(c);
        }
        return getLeft10();
    }

    public String cursorRight(int k) {
        for (int i = 0; i < k; ++i) {
            if (right.length() == 0) {
                break;
            }
            char c = right.charAt(right.length() - 1);
            right.deleteCharAt(right.length() - 1);
            left.append(c);
        }
        return getLeft10();
    }
}

/**
 * Your TextEditor object will be instantiated and called as such:
 * TextEditor obj = new TextEditor();
 * obj.addText(text);
 * int param_2 = obj.deleteText(k);
 * String param_3 = obj.cursorLeft(k);
 * String param_4 = obj.cursorRight(k);
 */
# @lc code=end

File: 2296.design-a-text-editor.go
#
# @lc app=leetcode id=2296 lang=golang
#
# [2296] Design a Text Editor
#

# @lc code=start
type TextEditor struct {
	left  []byte
	right []byte
}


func Constructor() TextEditor {
	return TextEditor{}
}


func (this *TextEditor) AddText(text string)  {
	this.left = append(this.left, []byte(text)...)
}


func (this *TextEditor) DeleteText(k int) int {
	n := len(this.left)
	del := k
	if del > n {
		del = n
	}
	this.left = this.left[:n-del]
	return del
}


func (this *TextEditor) CursorLeft(k int) string {
	n := len(this.left)
	moves := k
	if moves > n {
		moves = n
	}
	for i := 0; i < moves; i++ {
		ch := this.left[n-1]
		this.left = this.left[:n-1]
		this.right = append(this.right, ch)
		n--
	}
	l := len(this.left)
	start := 0
	if l > 10 {
		start = l - 10
	}
	return string(this.left[start:])
}


func (this *TextEditor) CursorRight(k int) string {
	n := len(this.right)
	moves := k
	if moves > n {
		moves = n
	}
	for i := 0; i < moves; i++ {
		ch := this.right[n-1]
		this.right = this.right[:n-1]
		this.left = append(this.left, ch)
		n--
	}
	l := len(this.left)
	start := 0
	if l > 10 {
		start = l - 10
	}
	return string(this.left[start:])
}


/**
 * Your TextEditor object will be instantiated and called as such:
 * obj := Constructor();
 * obj.AddText(text);
 * param_2 := obj.DeleteText(k);
 * param_3 := obj.CursorLeft(k);
 * param_4 := obj.CursorRight(k);
 */
# @lc code=end